chain/consensus/mirfbt/mine.go

```diff
@@ -0,0 +1,172 @@
+	"github.com/filecoin-project/lotus/chain/consensus/common"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+func Mine(ctx context.Context, miner address.Address, api v1api.FullNode) error {
+	addr := miner.String()
+	log.Infof("MirBFT miner %s: started", addr)
+	defer log.Info("MirBFT miner: stopped")
+
+	netName, nerr := api.StateNetworkName(ctx)
+	if nerr != nil {
+		return nerr
+	}
+	subnetID := address.SubnetID(netName)
+
+	log.Infof("MirBFT miner params: network name - %s, subnet ID - %s", netName, subnetID)
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debugf("MirBFT miner %s: context closed", addr)
+			return nil
+		default:
+			base, err := api.ChainHead(ctx)
+			if err != nil {
+				log.Errorw("failed to get the head of chain", "error", err)
+				continue
+			}
+
+			msgs, err := api.MpoolSelect(ctx, base.Key(), 1)
+			if err != nil {
+				log.Errorw("unable to select messages from mempool", "error", err)
+			}
+			log.Debugf("[subnet: %s, epoch: %d] retrieved %d msgs", subnetID, base.Height()+1, len(msgs))
+
+			crossMsgs, err := api.GetCrossMsgsPool(ctx, subnetID, base.Height()+1)
+			if err != nil {
+				log.Errorw("unable to select cross-messages from mempool", "error", err)
+			}
+			log.Debugf("[subnet: %s, epoch: %d] retrieved %d crossmsgs", subnetID, base.Height()+1, len(crossMsgs))
+
+			log.Infof("[subnet: %s, epoch: %d] try to create a block", subnetID, base.Height()+1)
+			bh, err := api.MinerCreateBlock(ctx, &lapi.BlockTemplate{
+				Miner:            miner,
+				Parents:          base.Key(),
+				BeaconValues:     nil,
+				Ticket:           nil,
+				Epoch:            base.Height() + 1,
+				Timestamp:        uint64(time.Now().Unix()),
+				WinningPoStProof: nil,
+				Messages:         msgs,
+				CrossMessages:    crossMsgs,
+			})
+			if err != nil {
+				log.Errorw("creating block failed", "error", err)
+				continue
+			}
+			if bh == nil {
+				log.Debug("created a nil block")
+				continue
+			}
+
+			log.Infof("[subnet: %s, epoch: %d] try to sync a block", subnetID, base.Height()+1)
+
+			err = api.SyncSubmitBlock(ctx, &types.BlockMsg{
+				Header:        bh.Header,
+				BlsMessages:   bh.BlsMessages,
+				SecpkMessages: bh.SecpkMessages,
+				CrossMessages: bh.CrossMessages,
+			})
+			if err != nil {
+				log.Errorw("unable to sync a block", "error", err)
+				continue
+			}
+
+			log.Infof("[subnet: %s, epoch: %d] mined a block %v", subnetID, bh.Header.Height, bh.Cid())
+		}
+	}
+}
+
+func (bft *MirBFT) CreateBlock(ctx context.Context, w lapi.Wallet, bt *lapi.BlockTemplate) (*types.FullBlock, error) {
+	log.Infof("create block for epoch %d", bt.Epoch)
+
+	for _, msg := range bt.Messages {
+		msgBytes, err := msg.Serialize()
+		if err != nil {
+			log.Error("unable to serialize message:", err)
+			continue
+		}
+
+		reqNo := mirbft.ReqNo(msg.Message.Nonce)
+		tx := common.NewSignedMessageBytes(msgBytes, nil)
+
+		err = bft.mir.Node.SubmitRequest(ctx, mirbft.ClientID(0), reqNo, tx, nil)
```

Comment:
Is there any relationship between `mirbft.ClientID` and the signing identity? (just a sanity-check, as I don't know how the mir-bft code works).

Reply:
Signing identity is a node identity, as far as I know. In future, client ID will be equal to Eudico peer ID